### PR TITLE
K63 animation effects

### PR DESCRIPTION
**Pull request by intxgo**
I have seen your pull request #627 draft, so I thought to add to it.

I have refactored a bit the cmd_hwanim function, hope you like it.

PS. I have no idea why GitHub shows so much changes. I have rebased commits from your branch on top of mine, so they shouldn't appear here.

In fact new commits are only

dec298e (already on ckb-next:master)
4184d0d

so you can just cherry pick them if the pull request does something silly.

### CODE REVIEW

File: src/daemon/led_hw.c
Comment (OrsanZ) [inline src/daemon/led_hw.c:17]: I don't see why we would need all these enums, I'd suggest removing them and sticking with plain ifs, instead of the switches.

File: src/daemon/led_hw.c
Comment (OrsanZ) [inline src/daemon/led_hw.c]: I would suggest keeping changes to a minimum here. I don't think we need the switches. Also, I think you've accidentally auto-formatted the `if()` to `if ()`, `for()` to `for ()`, etc. Please undo that.

File: src/daemon/led_hw.c
Comment (OrsanZ) [inline src/daemon/led_hw.c:188]: I really like this change. Well done for reverse engineering.

File: src/daemon/led_hw.c
Comment (OrsanZ) [inline src/daemon/led_hw.c]: What exactly does this do?

**Review by OrsanZ — CHANGES_REQUESTED** [review]

File: src/daemon/led_hw.c
Comment (intxgo) [reply src/daemon/led_hw.c:17]: Why to have comments and some magic numbers a few lines below? When you use enums, defines or constants it's self explanatory. In addition the number won't ever tell you what's the logical meaning of it.

Regarding the switches, if you really like, keep the snake ifs. Using switch with fall through is the cleanest way to validate if var fit into set of possible values. The main difference is that vertically you see them all, horizontally ugh... more than 3 is too much unless you have a big screen. But it's a matter of taste.

File: src/daemon/led_hw.c
Comment (intxgo) [reply src/daemon/led_hw.c]: Oh, I wasn't aware of any coding style guidelines

File: src/daemon/led_hw.c
Comment (intxgo) [reply src/daemon/led_hw.c]: hmm, good point, should be light_intensity_level

File: src/daemon/led_hw.c
Comment (OrsanZ) [reply src/daemon/led_hw.c]: There really aren't any, I just kinda do what I find in the code.

File: src/daemon/led_hw.c
Comment (OrsanZ) [reply src/daemon/led_hw.c:17]: I guess so. I would suggest adding comments above the ifs to explain what the numbers mean. I've tried to stick to the style that ckb-next has, not my own, and that seems to be how they do things. I'll remember the switch thing for my own future projects though :wink:

I'm sorry if I looked rude, I'm still not done answering which is why there is only the review.

Comment (intxgo) [reply src/daemon/led_hw.c:17]: No problem, I reverted the change for DARK CORE part

**Comment by OrsanZ** [comment]
Okay, I made a review. Thanks for contributing by the way, I wasn't expecting it. The review also isn't the most tidy since I have never done one before. Comments are above what's supposed to be changed, not below. :sweat_smile:

In terms of git, first of all I'd suggest undoing that merge commit using `git reset --hard HEAD~1`. I'll rebase onto the changed wireless branch myself and push it. Then you can fetch the wireless branch here and rebase your commit where you actually made changes on that.

The whole procedure would look like this:
`git reset --hard HEAD~1` be careful with this since it completely removes the last commit.
`git remote add orsanz https://github.com/OrsanZ/ckb-next.git`
`git fetch orsanz`
`git rebase orsanz/wireless`

That should fix the whole mess.
You can check if it looks good by doing:
`git log --oneline --graph`
That graph to the left is really helpful.

Give me a few minutes to rebase myself before you do this though.

Then you can take your commited changes and uncommit and unstage them using `git reset HEAD~1` again, make your changes and recommit, force-push to your fork, and that'll fix it.

I hope that this will help. If you haven't already, I really recommend reading at least chapter 2 and 3 of [the git book](https://git-scm.com/book/en/v2). And if you already knew all of this stuff, I'm not trying to be condescending, I just made the same mistakes a few days ago and the book really helped me.

**Comment by intxgo** [comment]
Oh, yes, I can rebase squash them. The point though is that when I do the pull request, git doesn't see how to merge my changes on top but asks for manual merge.

BTW, I could wait until you finish the pull request and merge it into ckb-next:wireless, but I just thought it might help if I fix the K63WL light right now.

**Comment by OrsanZ** [comment]
What my suggestion is trying to achieve is to get rid of the weird commits that come before yours. If you don't want to go through the troubles of that, I can just cherry pick your commits. I'd still suggest doing so since the PR is nicer that way but if you don't have the time or patience, I understand.

Also doing the PR is fine, I was gonna deal with the K63WL anyway.

**Comment by OrsanZ** [comment]
You figured it out! Nice.

**Comment by intxgo** [comment]
squashed, 
surprisingly now there's no merge needed, awesome!

**Comment by OrsanZ** [comment]
Thats what rebasing is for. :smile: 

One last thing before I merge... I hate to ask since you really love your switches, but could remove them and the enums since the implementation is inconsistent otherwise?

**Comment by intxgo** [comment]
Well, I would prefer to leave a self-readable code. That's why I also volunteered to refactor the part done by you. 
The commit bear reference to me, so don't worry about comments on the pull request to ckb-next:wireless, if the community really don't like my style I'll refactor it. Is it ok for you?

**Comment by OrsanZ** [comment]
You know what, if you still have the switch for the dark core, you can add them again since we then have a consistent implementation. If you accidentally deleted them, try recovering from the reflog. If that doesn't work, then I'll just do it and follow the same style.

**Comment by intxgo** [comment]
got it back
```
commit 3f2feb4c98c4b717ac77d81125bcfbee4805460e
Author: intrenalc <39905449+internalc@users.noreply.github.com>
Date:   Tue Aug 11 15:36:24 2020 +0200

    add K63WL animation effects

diff --git a/src/daemon/led_hw.c b/src/daemon/led_hw.c
index 3e21ddb..00154d5 100644
--- a/src/daemon/led_hw.c
+++ b/src/daemon/led_hw.c
@@ -2,16 +2,40 @@
 #include "usb.h"
 
 enum HWANIM_MODE {
-    HW_COLOURSHIFT = 0x00,
-    HW_COLOURPULSE = 0x01,
-    HW_RAINBOW     = 0x03,
-    HW_STATIC      = 0x07,
-    HW_NONE        = 0xff
+    HWANIM_COLOURSHIFT          = 0x00,
+    HWANIM_COLOURPULSE          = 0x01,
+    HWANIM_RAINBOW              = 0x03,
+    HWANIM_WAVE                 = 0x04,
+    HWANIM_VISOR                = 0x05,
+    HWANIM_RAIN                 = 0x06,
+    HWANIM_STATIC               = 0x07,
+    HWANIM_TYPELIGHTNING_KEY    = 0x08,
+    HWANIM_TYPELIGHTNING_RIPPLE = 0x09,
+    HWANIM_NONE                 = 0xff
 };
 
-int updatergb_wireless(usbdevice* kb, lighting* lastlight, lighting* newlight){
-    if(IS_DARK_CORE(kb)){
-        for(int i = 0; i < 3; i++) {
+enum HWANIM_ZONE_BIT {
+    HWANIM_ZONEBIT_FRONT = 0,
+    HWANIM_ZONEBIT_THUMB = 1,
+    HWANIM_ZONEBIT_REAR  = 2
+};
+
+enum HWANIM_DIRECTION {
+    HWANIM_DIRECTION_NONE  = 0,
+    HWANIM_DIRECTION_LEFT  = 1,
+    HWANIM_DIRECTION_RIGHT = 2
+};
+
+enum HWANIM_SPEED {
+    HWANIM_SPEED_NONE   = 0,
+    HWANIM_SPEED_LOW    = 1,
+    HWANIM_SPEED_MEDIUM = 2,
+    HWANIM_SPEED_HIGH   = 3,
+};
+
+int updatergb_wireless(usbdevice* kb, lighting* lastlight, lighting* newlight) {
+    if (IS_DARK_CORE(kb)) {
+        for (int i = 0; i < 3; i++) {
             char code[20] = "";
             snprintf(code, sizeof(code), "7:1:0:%02hhx%02hhx%02hhx", newlight->r[i], newlight->g[i], newlight->b[i]);
             cmd_hwanim(kb, 0, 0, 190 + i, code);
@@ -20,45 +44,77 @@ int updatergb_wireless(usbdevice* kb, lighting* lastlight, lighting* newlight){
     return 0;
 }
 
-void cmd_hwanim(usbdevice* kb, usbmode* dummy1, int dummy2, int keyindex, const char* code){
-    // Format: animation mode:speed:random:colour:colour2
-    uchar anim = 0xff, speed = 2, random = 0;
-    uchar r[2] = {}, g[2] = {}, b[2] = {};
-    sscanf(code, "%hhx:%hhu:%hhu:%2hhx%2hhx%2hhx:%2hhx%2hhx%2hhx", &anim, &speed, &random, r, g, b, r+1, g+1, b+1);
-
-    short zone = kb->keymap[keyindex].led;
-    // Limit values
-    if(anim != HW_COLOURSHIFT && anim != HW_COLOURPULSE && anim != HW_RAINBOW && anim != HW_STATIC && anim != HW_NONE)
-        return;
-    if(speed < 1 || speed > 3)
-        return;
-    if(random != 0 && random != 1)
-        return;
-    if(zone < 0 || zone > 2)
-        return;
-
-    // Deal with the random bytes' weirdness
-    if((anim == HW_COLOURSHIFT || anim == HW_COLOURPULSE) && !random)
-        random = 3;
-
-    // Dark Core zone bits:
-    // 1: Front
-    // 2: Thumb
-    // 4: Rear
-
-    // Dark Core commands:
-    // 00: Colour Shift
-    // 01: Colour Pulse
-    // 03: Rainbow
-    // 07: Static Colour
-    // FF: No animation (black)
-
-    if(IS_DARK_CORE(kb)) {
-        uchar data_pkt[MSG_SIZE] = {
+void cmd_hwanim(usbdevice* kb, usbmode* dummy1, int dummy2, int keyindex, const char* code) {
+    /* 
+       Unified format for the code string:
+       - DARK CORE Format: animation mode:speed:random:colour:colour2
+       - K63 WL    Format: animation mode:speed:dir
+    */
+
+    if (IS_DARK_CORE(kb)) {
+        // DARK CORE Format: animation mode:speed:random:colour:colour2
+        uchar anim = 0xff, speed = 2, random = 0;
+        uchar r[2] ={}, g[2] ={}, b[2] ={};
+        sscanf(code, "%hhx:%hhu:%hhu:%2hhx%2hhx%2hhx:%2hhx%2hhx%2hhx", &anim, &speed, &random, r, g, b, r+1, g+1, b+1);
+
+        short zone = kb->keymap[keyindex].led;
+
+        // Limit values
+
+        switch (anim) {
+        case HWANIM_COLOURSHIFT:
+        case HWANIM_COLOURPULSE:
+            // Deal with the random bytes' weirdness
+            if (!random) {
+                random = 3;
+            }
+            break;
+        case HWANIM_RAINBOW:
+        case HWANIM_STATIC:
+        case HWANIM_NONE:
+            break;
+        default:
+            return;
+        }
+
+        switch (speed) {
+        case HWANIM_SPEED_LOW:
+        case HWANIM_SPEED_MEDIUM:
+        case HWANIM_SPEED_HIGH:
+            break;
+        default:
+            return;
+        }
+
+        switch (zone) {
+        case HWANIM_ZONEBIT_FRONT:
+        case HWANIM_ZONEBIT_THUMB:
+        case HWANIM_ZONEBIT_REAR:
+            break;
+        default:
+            return;
+        }
+
+        if (random != 0 && random != 1)
+            return;
+
+        // Dark Core zone bits:
+        // 1: Front
+        // 2: Thumb
+        // 4: Rear
+
+        // Dark Core commands:
+        // 00: Colour Shift
+        // 01: Colour Pulse
+        // 03: Rainbow
+        // 07: Static Colour
+        // FF: No animation (black)
+
+        uchar data_pkt[MSG_SIZE] ={
             CMD_SET, 0xaa, 0
         };
 
-        if(anim == HW_STATIC)
+        if (anim == HWANIM_STATIC)
             speed = 0;
 
         data_pkt[4]  = 1 << zone;        // Bitmask.
@@ -75,26 +131,100 @@ void cmd_hwanim(usbdevice* kb, usbmode* dummy1, int dummy2, int keyindex, const
         data_pkt[14] = b[1];
         data_pkt[15] = 5 - zone;         // Layer byte
 
-        if(!usbsend(kb, data_pkt, 1))
+        if (!usbsend(kb, data_pkt, 1))
+            return;
+    }
+
+    if (IS_K63_WL(kb)) {
+        // K63 WL    Format: animation mode:speed:direction
+        uchar anim = HWANIM_NONE;
+        uchar speed = HWANIM_SPEED_NONE;
+        uchar direction = HWANIM_DIRECTION_NONE;
+        uchar opacity = 100; //100%
+        sscanf(code, "%hhx:%hhu:%hhu", &anim, &speed, &direction);
+
+        switch (anim) {
+        case HWANIM_COLOURSHIFT:
+        case HWANIM_COLOURPULSE:
+        case HWANIM_RAINBOW:
+        case HWANIM_VISOR:
+        case HWANIM_RAIN:
+        case HWANIM_TYPELIGHTNING_KEY:
+        case HWANIM_TYPELIGHTNING_RIPPLE:
+            switch (speed) {
+            case HWANIM_SPEED_LOW:
+            case HWANIM_SPEED_MEDIUM:
+            case HWANIM_SPEED_HIGH:
+                break;
+            default:
+                return;
+            }
+            //direction not supported
+            direction = HWANIM_DIRECTION_NONE;
+            break;
+        case HWANIM_STATIC:
+        case HWANIM_NONE:
+            //direction and speed not supported
+            direction = HWANIM_DIRECTION_NONE;
+            speed = HWANIM_SPEED_NONE;
+            break;
+        case HWANIM_WAVE:
+            switch (direction) {
+            case HWANIM_DIRECTION_LEFT:
+            case HWANIM_DIRECTION_RIGHT:
+                break;
+            default:
+                return;
+            }
+            break;
+        default:
             return;
-    } else if (IS_K63_WL(kb)) {
-        uchar data_pkt[MSG_SIZE] = {
+        }
+
+        uchar data_pkt[MSG_SIZE] ={
             CMD_SET, 0xaa, 0
         };
 
-        data_pkt[4]  = 0;
-        data_pkt[5]  = anim;
-        data_pkt[8]  = 100;
-        data_pkt[9]  = r[0];
-        data_pkt[10] = g[0];
-        data_pkt[11] = b[0];
-        // Second color for effects like alternating color shift
-        data_pkt[12] = r[0];
-        data_pkt[13] = g[0];
-        data_pkt[14] = b[0];
-        data_pkt[15] = 0;
+        //iCUE always changes the animation in 2 steps, so let's do the same
+
+        data_pkt[3]  = 0;
+        data_pkt[4]  = HWANIM_NONE;
+        data_pkt[5]  = HWANIM_SPEED_NONE;
+        data_pkt[6]  = HWANIM_DIRECTION_NONE;
+        data_pkt[7]  = 100;
+
+        if (!usbsend(kb, data_pkt, 1))
+            return -1;
+
+        data_pkt[3]  = 0;
+        data_pkt[4]  = anim;
+        data_pkt[5]  = speed;
+        data_pkt[6]  = direction;
+        data_pkt[7]  = 100;
+
+        if (!usbsend(kb, data_pkt, 1))
+            return -1;
+
+        //Optional:
+
+        //Turn on the light to ensure the animation is visible
+        //TODO: support it via the common way (how?)
+
+        /*
+            What's interesting, the hardware button for backlight intensity
+            on K63WL just sends the keypress code to Windows, then the iCUE
+            responds with change light command below.
+        */
+
+        uchar level = 1; //1,2,3 where l3 can be globally disabled
+
+        memset(data_pkt, 0, MSG_SIZE);
+        data_pkt[1]  = 5;
+        data_pkt[2]  = 2;
+        data_pkt[3]  = 0;
+        data_pkt[4]  = level;
 
-        if(!usbsend(kb, data_pkt, 1))
+        if (!usbsend(kb, data_pkt, 1))
             return -1;
     }
     return 0;
diff --git a/src/daemon/usb.h b/src/daemon/usb.h
index 69304cd..772f63e 100644
--- a/src/daemon/usb.h
+++ b/src/daemon/usb.h
@@ -217,7 +217,7 @@ const char* product_str(ushort product);
 #define IS_DETACHABLE(kb)               (IS_K63_WL(kb) || IS_DARK_CORE(kb))
 
 /// Used for ID combinations which support wireless.
-#define IS_DETACHABLE_ID(vendor, product)((vendor) == V_CORSAIR && (product == P_DARK_CORE || (product) == P_DARK_CORE_WL || (product) == P_DARK_CORE_SE || (product) == P_DARK_CORE_SE_WL || (product) == P_K63_NRGB_WL || (product) == P_K63_NRGB_WL2))
+#define IS_DETACHABLE_ID(vendor, product)((vendor) == V_CORSAIR && (product == P_DARK_CORE || (product) == P_DARK_CORE_WL || (product) == P_DARK_CORE_SE || (product) == P_DARK_CORE_SE_WL || (product) == P_K63_NRGB_WL || (product) == P_K63_NRGB_WL2 || (product) == P_K63_NRGB_WL3 || (product) == P_K63_NRGB_WL4))
 
 /// Used for devices which are wireless.
 #define IS_WIRELESS(kb)                 ((kb)->product == P_DARK_CORE_WL || (kb)->product == P_DARK_CORE_SE_WL || IS_K63_WL(kb))
```

**Comment by OrsanZ** [comment]
Nice. Now just add the dark core part and amend it using `git commit --amend --no-edit`.
This effectively creates a new commit and squashes it with the last.

**Comment by intxgo** [comment]
Done.
BTW, I've sent my keyboard for RMA replacement. I hope all this get's merged before I got it back. Then I'll see what's left to be done. 
Thanks for all  your contributions anyway!

**Comment by OrsanZ** [comment]
Merged. By the way, I messed up and accidentally put a non-noreply email in the CueShark repository. It's fixed in the main repo but it's still in your fork. Could you please delete the fork, since your changes were already merged? I don't want it getting added to spam lists and the like.

**Comment by intxgo** [comment]
> Merged. By the way, I messed up and accidentally put a non-noreply email in the CueShark repository. It's fixed in the main repo but it's still in your fork. Could you please delete the fork, since your changes were already merged? I don't want it getting added to spam lists and the like.

sure

**Comment by OrsanZ** [comment]
Thanks. By the way, I didn't mean to request a review :sweat_smile:

**Pull request merged**
